iOS Code Refactor and New Tests for Image Classifier

priankakariatyml opened the pull request:
1. Fixed review comments
2. Added new tests for Image Classifier
3. Refactored code

priankakariatyml commented on tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m:
```diff
@@ -100,6 +100,12 @@ - (nullable TFLClassificationResult *)classifyWithGMLImage:(GMLImage *)image
+  if (image == nil) {
+    if (error != nil) {
+      *error = [TFLCommonUtils customErrorWithCode:TFLSupportErrorCodeInvalidArgumentError
+                                       description:@"GMLImage argument cannot be nil."];
```
@khanhlvg Long back we had a discussion that the error null check here should be incorporated into `[TFLCommonUtils customErrorWithCode:]` in order to remove redundant error check every time we have to create an error. I had made this change but seems to have been reversed in a later commit while some PR was merged. I am not sure if this was accidental. If it was, can I incorporate the null error check into the method in a subsequent PR?

khanhlvg replied:
Yes please incorporate the null check into TFLCommonUtils.

One thing that can cause the reverse is that we import the PR to merge internally, not on GitHub. Therefore, if you add a commit to a PR after we initiated the PR import process, that commit will end up not being included in the merge.

priankakariatyml replied:
Okay. I will

khanhlvg replied:
Please let me know once you've added this refactor. The rest of the code LGTM so once that's done, I'll import this PR for internal review.

priankakariatyml replied:
This will bloat up the PR and change a lot of files other than image classifier since the method calls are in a lot of places. But they are minor changes. Is that okay?

priankakariatyml replied:
Thought I would do it in a follow up PR.

priankakariatyml replied:
I have incorporated the error refractor. LMK if it's a lot of code.

khanhlvg commented on tensorflow_lite_support/ios/task/processor/sources/TFLClassificationResult+Helpers.m:
```diff
@@ -24,21 +24,15 @@ + (TFLClassificationResult *)classificationResultWithCResult:
-    classificationHead.categories = classes;
-    classificationHead.headIndex = i;
-    [classificationHeads addObject:classificationHead];
+    [classificationHeads addObject:[[TFLClassifications alloc] initWithHeadIndex:i
```
nit: You should keep an intermediate variable TFLClassifications *classificationHead to improve readability.

priankakariatyml replied:
> @priankakariatyml What's the source of the burger image? Is it released under a license such as CC0 that allows us to add it into the repo?

I have added an alpha channel of 1s to the existing burger.jpg in the repo using a python script. So it its okay.

khanhlvg commented:
@priankakariatyml What's the source of the burger image? Is it released under a license such as CC0 that allows us to add it  into the repo?

khanhlvg commented:
@priankakariatyml Can you move the tests into a different PR? This PR is quite large so we may have an issue finding internal reviewer.

priankakariatyml commented:
> @priankakariatyml Can you move the tests into a different PR? This PR is quite large so we may have an issue finding internal reviewer.

I have removed the tests for the time being. Test files still have minor changes fixing the code review comments from internal review.

priankakariatyml commented:
> @priankakariatyml What's the source of the burger image? Is it released under a license such as CC0 that allows us to add it into the repo?

I have added an alpha channel of 1s to the existing burger.jpg in the repo using a python script. So it its okay. For the time being I have removed it. Will add it back with the new tests.

khanhlvg commented on tensorflow_lite_support/ios/sources/TFLCommonUtils.h:
```diff
@@ -26,31 +26,33 @@ NS_ASSUME_NONNULL_BEGIN
+ * no error will be saved.
  */
-+ (NSError *)errorWithCError:(TfLiteSupportError *)supportError;
++ (void)errorWithCError:(TfLiteSupportError *)supportError error:(NSError **)error;
```
ditto.
I'd suggest:
```
+ (void)convertCError:(TfLiteSupportError *)supportError toError:(NSError **)error;
```

priankakariatyml commented on tensorflow_lite_support/ios/sources/TFLCommonUtils.h:
```diff
@@ -26,31 +26,33 @@ NS_ASSUME_NONNULL_BEGIN
-+ (NSError *)customErrorWithCode:(NSInteger)code
-                description:(NSString *)description;
-
++ (void)customErrorWithCode:(NSInteger)code
```
SGTM. I have made this change.

priankakariatyml commented on tensorflow_lite_support/ios/task/processor/sources/TFLClassificationOptions+Helpers.m:
```diff
@@ -20,33 +20,24 @@ @implementation TFLClassificationOptions (Helpers)
     return nil;
   }
 
+  char **cStrings = (char **)[TFLCommonUtils mallocWithSize:strings.count * sizeof(char *)
```
Fixed

priankakariatyml commented on tensorflow_lite_support/ios/task/processor/sources/TFLClassificationOptions+Helpers.m:
```diff
@@ -84,8 +75,8 @@ - (BOOL)copyToCOptions:(TfLiteClassificationOptions *)cClassificationOptions
-  if (self.displayNamesLocal) {
-    cClassificationOptions->display_names_local = (char *)self.displayNamesLocal.UTF8String;
+  if (self.displayNamesLocale) {
+    cClassificationOptions->display_names_local = (char *)self.displayNamesLocale.UTF8String;
```
Fixed

priankakariatyml commented on tensorflow_lite_support/ios/sources/TFLCommonUtils.m:
```diff
@@ -20,30 +20,33 @@
+    [TFLCommonUtils createCustomError:error
+                             withCode:TFLSupportErrorCodeInvalidArgumentError
+                          description:@"Invalid memory size passed for allocation of object."];
+    return nil;
```
Fixed

priankakariatyml commented on tensorflow_lite_support/ios/task/processor/sources/TFLCategory.h:
```diff
@@ -19,18 +19,38 @@ NS_ASSUME_NONNULL_BEGIN
+ * @return An instance of TFLCategory initialized to
+ * the specified values.
+ */
+- (instancetype)initWithIndex:(NSInteger)index
```
Fixed

priankakariatyml commented on tensorflow_lite_support/ios/task/processor/sources/TFLCategory.m:
```diff
@@ -14,9 +14,25 @@
-@synthesize label;
-@synthesize score;
-@synthesize classIndex;
+@implementation TFLCategory {
```
Fixed

khanhlvg commented:
Merged.
